tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+```go
+  // Read the radius-config.yaml file
+  configFile := filepath.Join(tmpDir, "radius-config.yaml")
+  config, err := ioutil.ReadFile(configFile)
+  if err != nil {
+    logger.Error(err, "unable to read radius-config.yaml file")
+    return ctrl.Result{}, err
+  }
+
+  // Parse the radius-config.yaml file
+  var radiusConfig RadiusConfig
+  if err := yaml.Unmarshal(config, &radiusConfig); err != nil {
+    logger.Error(err, "unable to parse radius-config.yaml file")
+    return ctrl.Result{}, err
+  }
+```
+
+The `radius-config.yaml` file will have the following structure:
+> Note: The structure of the `radius-config.yaml` file is subject to change. We should discuss this in the design meeting. I considered adapting the existing `.rad/config.yaml` file format, but there were too many differences between the two formats, so I decided to create a new format specifically for the Radius Flux Controller.
+```yaml
+config:
+  bicep:
+    - file: main.bicep
+      parameters: main.bicepparam
+      namespace: default
+    - file: secondary.bicep
+  providers:
+    - radius:
+        scope: /planes/radius/local/resourcegroups/default
+    - aws:
+        scope: /planes/aws/aws/accounts/000000000/regions/us-west-2
+    - azure:
+        scope: 	/planes/azure/azurecloud/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/my-group
+```
+
+#### Compiling the Bicep File
+
+The controller will compile the Bicep file using the Bicep CLI, which is installed on the controller. It will use existing functionality to leverage the Bicep compiler: https://github.com/radius-project/radius/blob/87f7ea1b8d814a266fb89333ea4e5b38463ed7f9/pkg/cli/bicep/types.go#L46
+
+#### Creating the `DeploymentTemplate` Resource
+
+The controller will create (or update) a `DeploymentTemplate` resource on the cluster with the compiled Bicep file as the template:
+```go
+  // Create the DeploymentTemplate resource
+  deploymentTemplate := &radappiov1alpha3.DeploymentTemplate{
+    ObjectMeta: metav1.ObjectMeta{
+      Name:      radiusConfig.Name,
+      Namespace: radiusConfig.Namespace,
+    },
+    Spec: radappiov1alpha3.DeploymentTemplateSpec{
+      Template: string(compiledBicep),
+    },
+  }
+
+  // Create or update the DeploymentTemplate resource
+  if err := r.CreateOrUpdate(ctx, deploymentTemplate); err != nil {
+    logger.Error(err, "unable to create or update DeploymentTemplate resource")
+    return ctrl.Result{}, err
+  }
+```
+
+#### Deleting Stale `DeploymentTemplate` Resources
+
+The controller will list all `DeploymentTemplate` resources in the cluster that have been created from this Git repository and delete any resources that are not present in the Git repository:
+```go
+	// List all DeploymentTemplates on the cluster that are from the same git repository
```

Comment:
Are we talking about older versions of a DeploymentTemplate for a given repository?

Reply:
Can a repository have multiple DeploymentTemplates?

Reply:
Yes, it can have multiple.